daffodil-runtime1/src/main/scala-2.11/org/apache/daffodil/infoset/DataValue.scala

```diff
@@ -99,8 +99,8 @@ final class DataValue[+T <: AnyRef, +X <: AnyRef] private (val v: T) extends Any
   @inline def getString = v.asInstanceOf[JString]
   @inline def getURI = v.asInstanceOf[URI]
 
-  @inline def getNonNullable = this.asInstanceOf[DataValue[T, X with NonNullable]]
-  @inline def getNullablePrimitive = this.asInstanceOf[DataValue.DataValuePrimitiveNullable]
+  @inline def getNonNullable:DataValue[T, X with NonNullable] = new DataValue(v)
+  @inline def getNullablePrimitive:DataValue.DataValuePrimitiveNullable = new DataValue(v)
```

Comment:
Not actually relevent, but using asInstanceOf above introduces boxing.

Reply:
Really. Wow, so you can't even asInstanceOf an AnyVal type without it being boxed....

We need a section in our wiki about using AnyVal types in the coding style and coding for performance things. In code review we should be looking for this sort of thing. Right now I think it just says there are some non-boxing and value class types like MStack and suggests using them, but doesn't give a do's and don'ts list.

Reply:
Yeah. I did a grep through a dissasembly of our source code, and found 188 instances of Maybe."<init>" and 78 instances of DataValue."<init>"

One place these come up seem to be pass-by-name parameters, which would probably account for alot of these. I didn't look through all of them, so I do not know if there are other places we introduced boxing.

I did verify that all of the toDataValue functions do not introduce boxing, so we are doing at least something write.

I googled around and found a recent Pre-SIP to introduce true non boxing wrapper types: https://contributors.scala-lang.org/t/pre-sip-unboxed-wrapper-types/987

It is worth keeping an eye on this and switching over to it when/if it becomes available.